src/Target.php

```diff
@@ -92,292 +50,332 @@ abstract class Target
-     * @var string The date format for the log timestamp.
-     * Defaults to Y-m-d H:i:s.u
+     * @var string The date format for the log timestamp. Defaults to `Y-m-d H:i:s.u`.
      */
     private string $timestampFormat = 'Y-m-d H:i:s.u';
 
-
     /**
-     * @var bool|callable
+     * @var bool|callable Enables or disables the current target to export.
      */
     private $enabled = true;
 
     /**
-     * Exports log {@see Target::$messages} to a specific destination.
+     * Exports log messages to a specific destination.
      * Child classes must implement this method.
      */
-    abstract public function export(): void;
+    abstract protected function export(): void;
+
+    /**
+     * When defining a constructor in child classes, you must call `parent::__construct()`.
+     */
+    public function __construct()
+    {
+        $this->categories = new MessageCategory();
+        $this->messages = new MessageCollection();
+    }
 
     /**
      * Processes the given log messages.
-     * This method will filter the given messages with {@see Target::$levels} and {@see Target::$categories}.
+     *
+     * This method will filter the given messages with levels and categories.
+     * The message structure follows that in {@see MessageCollection::$messages}.
      * And if requested, it will also export the filtering result to specific medium (e.g. email).
      *
-     * @param array $messages log messages to be processed. See {@see Logger::$messages} for the structure
-     * of each message.
-     * @param bool $final whether this method is called at the end of the current application
+     * @param array $messages Log messages to be processed.
+     * @param bool $final Whether this method is called at the end of the current application.
      */
     public function collect(array $messages, bool $final): void
     {
-        $this->messages = array_merge(
-            $this->messages,
-            $this->filterMessages($messages, $this->levels, $this->categories, $this->except)
-        );
+        $this->messages->addMultiple($this->filterMessages($messages));
+        $count = $this->messages->count();
 
-        $count = count($this->messages);
         if ($count > 0 && ($final || ($this->exportInterval > 0 && $count >= $this->exportInterval))) {
-            if (($context = $this->getContextMessage()) !== '') {
-                $this->messages[] = [
-                    LogLevel::INFO,
-                    $context,
-                    [
-                        'category' => static::DEFAULT_CATEGORY,
-                        'time' => $_SERVER['REQUEST_TIME_FLOAT'] ?? \microtime(true),
-                    ],
-                ];
+            if (($contextMessage = $this->getContextMessage()) !== '') {
+                $this->messages->add(LogLevel::INFO, $contextMessage, [
+                    'category' => MessageCategory::DEFAULT,
+                    'time' => $_SERVER['REQUEST_TIME_FLOAT'] ?? microtime(true),
```

Comment:
That looks suspicious. ` $_SERVER['REQUEST_TIME_FLOAT']` is the time of the request beginning.

Reply:
Oops! Forgot to delete.